### PR TITLE
scheduler use spin::Mutex

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Signed-off-by: zhangjinpeng1987 <zhangjinpeng@pingcap.com>

## What have you changed? (mandatory)

Replace Mutex with Spinlock in scheduler.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

- [x] micro benchmark

```
#[bench]
fn bench_spinlock(b: &mut Bencher) {
    let spinlock = Spinlock::new(0);
    b.iter(|| {
        let _x = spinlock.lock();
    });
}

#[bench]
fn bench_mutex(b: &mut Bencher) {
    let m = Mutex::new(0);
    b.iter(|| {
        let _x = m.lock().unwrap();
    });
}

jinpengdeMacBook-Pro:bench-spinlock-mutex jinpengzhang$ cargo bench
    Finished release [optimized] target(s) in 0.07s
     Running target/release/deps/bench_spinlock_mutex-b6f6145e45142f47

running 2 tests
test bench_mutex    ... bench:          28 ns/iter (+/- 1)
test bench_spinlock ... bench:          13 ns/iter (+/- 1)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out
```

- [x] benchmark

Master run sysbench insert workload for 400 seconds with 256 concurrency:
sched-mutex

After replace Mutex with Spinlock, the related CPU time has been reduced by about 10s:
sched-spinlock

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

**Comment by siddontang** [comment]
em, any big difference? 

seem Mutex may still enter Spin phase at first.

**Comment by zhangjinpeng1987** [comment]
As the micro benchmark shows the spinlock(https://docs.rs/spinlock/0.1.2/src/spinlock/lib.rs.html#1-191) is more light.

```
#[bench]
fn bench_spinlock(b: &mut Bencher) {
    let spinlock = Spinlock::new(0);
    b.iter(|| {
        let _x = spinlock.lock();
    });
}

#[bench]
fn bench_mutex(b: &mut Bencher) {
    let m = Mutex::new(0);
    b.iter(|| {
        let _x = m.lock().unwrap();
    });
}

jinpengdeMacBook-Pro:bench-spinlock-mutex jinpengzhang$ cargo bench
    Finished release [optimized] target(s) in 0.07s
     Running target/release/deps/bench_spinlock_mutex-b6f6145e45142f47

running 2 tests
test bench_mutex    ... bench:          28 ns/iter (+/- 1)
test bench_spinlock ... bench:          13 ns/iter (+/- 1)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out
```

**Comment by breeswish** [comment]
LGTM. But spinlock is a deprecated crate. [`spin`](https://github.com/mvdnes/spin-rs) should be used instead.

**Comment by kennytm** [comment]
Could you compare with the Mutex from `parking_lot`?

**Comment by zhangjinpeng1987** [comment]
I bench std::sync::Mutex, spinlock::Spinlock, spin::Mutex and parking_lot::Mutex, on CentOS the result is 
```
running 4 tests
test bench_mutex       ... bench:          25 ns/iter (+/- 1)
test bench_parking_lot ... bench:          15 ns/iter (+/- 0)
test bench_spin_mutex  ... bench:          14 ns/iter (+/- 0)
test bench_spinlock    ... bench:          15 ns/iter (+/- 0)
```
On MacOS
```
test bench_mutex       ... bench:          29 ns/iter (+/- 16)
test bench_parking_lot ... bench:          13 ns/iter (+/- 2)
test bench_spin_mutex  ... bench:           9 ns/iter (+/- 1)
test bench_spinlock    ... bench:          13 ns/iter (+/- 2)
```
The spin::Mutex performs better than the others.

File: src/storage/txn/latch.rs [hunk 1]
```diff
@@ -3,9 +3,10 @@
 use std::collections::hash_map::DefaultHasher;
 use std::collections::VecDeque;
 use std::hash::{Hash, Hasher};
-use std::sync::Mutex;
 use std::usize;
 
+use spin;
```
Comment (breeswish) [inline src/storage/txn/latch.rs]: can be removed

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -20,9 +20,10 @@
 //! is ensured by the transaction protocol implemented in the client library, which is transparent
 //! to the scheduler.
 
+use spin;
```
Comment (breeswish) [inline src/storage/txn/scheduler.rs]: can be removed

Comment (siddontang) [reply src/storage/txn/scheduler.rs]: I prefer use spin::Mutex here directly.

**Comment by zhangjinpeng1987** [comment]
@breeswish @siddontang PTAL again.

**Pull request merged**